### ISSUE TITLE
dvc: drop python 3.7

### ISSUE THREAD

**Issue by efiop** [iterative/dvc#7708]
3.7 EOL is about a year away, but we already see projects that we depend on start to drop it. E.g. https://github.com/fsspec/adlfs/pull/307 And we will see much more in the upcoming months (e.g. https://numpy.org/neps/nep-0029-deprecation_policy.html).

We went through the process of keeping 3.6 alive a year ago and that was a pretty big waste of time, hunting for deprecations in our dependencies and contributing patches back just to drop 3.6 again less then a year later.

<img width="1461" alt="Screenshot 2022-05-06 at 13 56 11" src="https://user-images.githubusercontent.com/5367102/167118818-0c13b7b2-869f-4cb6-a7bc-0b7ddb233e7c.png">

Looking at our pypi stats, 3.7 downloads while significant, are still less than 30% than the newer versions and the only case where we care about that is system-wide python installs, that we don't recommend (we recommend venv instead) and the rest of the packages like deb/exe/snap/etc all come with their own python interpreter.

Considering all of these factors, I think we should drop 3.7 ASAP and focus on stuff that matters.

**Comment by skshetry** [comment]
Also, note that Ubuntu 20.04 comes with Python 3.8.

**Comment by shcheklein** [comment]
Are there any insights on where do these downloads come from? It's very tempting :) The only thing we need to check if it becomes a blocker in some large orgs (for whatever reason some large org has Python 3.7 still locked), or getting it from some pre-backed Dockers that would take them a lot of time to migrate. Would be sad to realize that it can become a blocker in some deals.

**Comment by efiop** [comment]
@shcheklein No insights about that, unfortunately. We've been recommending all-in-one packages in https://dvc.org/doc/install for a long time now, the only potential issue is linux + venv + python3.7, but it is easy to use deb/rpm/snap/etc instead. If someone absolutely has to use dvc as a library with python3.7, they can use older dvc version until they are ready to migrate. And likely they would already have some other library in their toolbelt drop 3.7 (e.g. numpy ecosystem). We've been dropping python versions for many years now, and even though this concern is raised every time, it has never been a significant issue so far.

**Comment by shcheklein** [comment]
Yep, agreed, we've never done this one year prior EOL though :). On the other hand situation is different - it's not even close as dropping Python 2 one year before. 3.x adoption is moving way faster.

**Comment by tirkarthi** [comment]
Ending 3.7 support might also help with pickle related errors caused due to mismatch of Python versions since Python 3.8 uses protocol 5 through https://peps.python.org/pep-0574/.

Related issue : https://github.com/iterative/dvc/issues/7726

**Comment by samhodge-aiml** [comment]
I am not sure why you removed all the existing versions from Python Package Index.

I am no longer able to run old projects.

I am not sure what experience you have with stable productions, but this is really unsuitable.

**Comment by skshetry** [comment]
> I am not sure why you removed all the existing versions from Python Package Index.

@samhodge-aiml, what do you mean? You can install all versions from PyPI, we haven't removed anything.

**Comment by samhodge-aiml** [comment]
Show me

```
$ python --version
Python 3.7.12
$ pip install dvc-gs==2.9.4
Python >=3.8; 2.19.1 Requires-Python >=3.8; 2.20.0 Requires-Python >=3.8; 2.21.0 Requires-Python >=3.8; 2.21.1 Requires-Python >=3.8; 2.22.0 Requires-Python >=3.8
ERROR: Could not find a version that satisfies the requirement dvc-gs==2.9.4 (from versions: none)
ERROR: No matching distribution found for dvc-gs==2.9.4
```

**Comment by skshetry** [comment]
You can install `dvc` with `gs` using:
```console
pip install 'dvc[gs]'
```

Note that we have stopped supporting Python 3.7, and it is getting EOL'd in 2 months.
So you can only install `dvc<=2.11.0`.

**Comment by samhodge-aiml** [comment]
Most of the Hollywood visual effects industry is using RHEL CentOS 7

Because something is EOL, it doesnt stop you from using it.

**Comment by skshetry** [comment]
You can use dvc up to 2.11.0. We dropped support after that. You can provide extras inside brackets, like `dvc[s3]`/`dvc[gs]`, etc.

**Comment by samhodge-aiml** [comment]
yeah thanks for the heads up

our existing pipeline was tuned into the `dvc-gs` alias I think it might have been via a `conda` indirection which stopped working

Its just I have enough on my plate without things breaking that I haven't touched.

Code rot happens....

**Comment by skshetry** [comment]
yes, in conda, you can install it as `conda install dvc-gs`.
